File: packages/runtime/src/directory.ts
import * as tg from "./index.ts";

export let directory = async (
	...args: Array<tg.Unresolved<tg.Directory.Arg>>
) => {
	return await tg.Directory.new(...args);
};

export class Directory {
	#state: tg.Directory.State;

	constructor(state: tg.Directory.State) {
		this.#state = state;
	}

	get state(): tg.Directory.State {
		return this.#state;
	}

	static withId(id: tg.Directory.Id): tg.Directory {
		return new tg.Directory({ id, stored: true });
	}

	static withObject(object: tg.Directory.Object): tg.Directory {
		return new tg.Directory({ object, stored: false });
	}

	static fromData(data: tg.Directory.Data): tg.Directory {
		return tg.Directory.withObject(tg.Directory.Object.fromData(data));
	}

	static async new(
		...args: Array<tg.Unresolved<tg.Directory.Arg>>
	): Promise<tg.Directory> {
		if (args.length === 1) {
			let arg = await tg.resolve(args[0]);
			if (
				typeof arg === "object" &&
				"node" in arg &&
				typeof arg.node === "number"
			) {
				return tg.Directory.withObject(arg as tg.Graph.Reference);
			}
		}
		let resolved = await Promise.all(args.map(tg.resolve));
		let entries = await resolved.reduce<
			Promise<{ [key: string]: tg.Graph.Edge<tg.Artifact> }>
		>(async function reduce(promise, arg) {
			let entries = await promise;
			if (arg === undefined) {
				// If the arg is undefined, then continue.
			} else if (arg instanceof tg.Directory) {
				// If the arg is a directory, then apply each entry.
				for (let [name, entry] of Object.entries(await arg.entries())) {
					// Get an existing entry.
					let existingEntry = entries[name];

					// Merge the existing entry with the entry if they are both directories.
					if (
						existingEntry instanceof tg.Directory &&
						entry instanceof tg.Directory
					) {
						entry = await tg.Directory.new(existingEntry, entry);
					}

					// Set the entry.
					entries[name] = entry;
				}
			} else if (typeof arg === "object") {
				// If the arg is an object, then apply each entry.
				for (let [key, value] of Object.entries(arg)) {
					// Separate the first normal path component from the trailing path components.
					let [firstComponent, ...trailingComponents] = tg.path.components(key);
					if (firstComponent === undefined) {
						throw new Error("the path must have at least one component");
					}
					if (!tg.path.Component.isNormal(firstComponent)) {
						throw new Error("all path components must be normal");
					}
					let name = firstComponent;

					// Get an existing entry.
					let existingEntry = entries[name];

					// Remove the entry if it is not a directory.
					if (!(existingEntry instanceof tg.Directory)) {
						existingEntry = undefined;
					}

					if (trailingComponents.length > 0) {
						// If there are trailing path components, then recurse.
						let trailingPath = tg.path.fromComponents(trailingComponents);

						// Merge the entry with the trailing path.
						let newEntry = await tg.Directory.new(existingEntry, {
							[trailingPath]: value,
						});

						// Add the entry.
						entries[name] = newEntry;
					} else {
						// If there are no trailing path components, then create the artifact specified by the value.
						if (value === undefined) {
							delete entries[name];
						} else if (typeof value === "number") {
							entries[name] = { node: value };
						} else if (
							typeof value === "object" &&
							typeof value.node === "number"
						) {
							entries[name] = value;
						} else if (
							typeof value === "string" ||
							value instanceof Uint8Array ||
							value instanceof tg.Blob
						) {
							let newEntry = await tg.file(value);
							entries[name] = newEntry;
						} else if (
							value instanceof tg.File ||
							value instanceof tg.Symlink
						) {
							entries[name] = value;
						} else {
							entries[name] = await Directory.new(existingEntry, value);
						}
					}
				}
			} else {
				return tg.unreachable();
			}
			return entries;
		}, Promise.resolve({}));
		return Directory.withObject({ entries });
	}

	static expect(value: unknown): Directory {
		tg.assert(value instanceof Directory);
		return value;
	}

	static assert(value: unknown): asserts value is Directory {
		tg.assert(value instanceof Directory);
	}

	get id(): Directory.Id {
		if (this.#state.id! !== undefined) {
			return this.#state.id;
		}
		let object = this.#state.object!;
		let data = Directory.Object.toData(object);
		let id = syscall("object_id", { kind: "directory", value: data });
		this.#state.id = id;
		return id;
	}

	async object(): Promise<Directory.Object> {
		await this.load();
		return this.#state.object!;
	}

	async load(): Promise<tg.Directory.Object> {
		if (this.#state.object === undefined) {
			let data = await syscall("object_get", this.#state.id!);
			tg.assert(data.kind === "directory");
			let object = Directory.Object.fromData(data.value);
			this.#state.object = object;
		}
		return this.#state.object!;
	}

	async store(): Promise<tg.Directory.Id> {
		await tg.Value.store(this);
		return this.id;
	}

	async children(): Promise<Array<tg.Object>> {
		let object = await this.load();
		return tg.Directory.Object.children(object);
	}

	async get(arg: string): Promise<tg.Artifact> {
		let artifact = await this.tryGet(arg);
		tg.assert(artifact, `Failed to get the directory entry "${arg}".`);
		return artifact;
	}

	async tryGet(arg: string): Promise<tg.Artifact | undefined> {
		let components = tg.path.components(arg);
		let artifact: tg.Artifact = this;
		let parents: Array<tg.Directory> = [];
		while (true) {
			let component = components.shift();
			if (component === undefined) {
				break;
			} else if (component === tg.path.Component.Root) {
				throw new Error("invalid path");
			} else if (component === ".") {
				continue;
			} else if (component === "..") {
				let parent = parents.pop();
				if (!parent) {
					throw new Error("path is external");
				}
				artifact = parent;
				continue;
			}
			if (!(artifact instanceof Directory)) {
				return undefined;
			}
			let entries = await artifact.entries();
			let entry: tg.Artifact | undefined = entries[component];
			if (entry === undefined) {
				return undefined;
			}
			parents.push(artifact);
			artifact = entry;
			if (entry instanceof tg.Symlink) {
				let artifact_ = await entry.artifact();
				let path_ = await entry.path();
				if (artifact_ === undefined && path_ !== undefined) {
					let parent = parents.pop();
					if (!parent) {
						throw new Error("path is external");
					}
					artifact = parent;
					components.unshift(...tg.path.components(path_));
				} else if (artifact_ !== undefined && path_ === undefined) {
					return artifact_;
				} else if (artifact_ instanceof tg.Directory && path_ !== undefined) {
					return await artifact_.tryGet(path_);
				} else {
					throw new Error("invalid symlink");
				}
			}
		}
		return artifact;
	}

	async entries(): Promise<{ [key: string]: tg.Artifact }> {
		let entries: { [key: string]: tg.Artifact } = {};
		for await (let [name, artifact] of this) {
			entries[name] = artifact;
		}
		return entries;
	}

	async *walk(): AsyncIterableIterator<[string, tg.Artifact]> {
		for await (let [name, artifact] of this) {
			yield [name, artifact];
			if (artifact instanceof Directory) {
				for await (let [entryName, entryArtifact] of artifact.walk()) {
					yield [tg.path.join(name, entryName), entryArtifact];
				}
			}
		}
	}

	async *[Symbol.asyncIterator](): AsyncIterator<[string, tg.Artifact]> {
		let object = await this.object();
		let entries: { [key: string]: tg.Artifact } | undefined;
		if ("entries" in object) {
			entries = Object.fromEntries(
				await Promise.all(
					Object.entries(object.entries).map(async ([name, edge]) => {
						tg.assert(typeof edge === "object", "expected an obejct");
						if ("node" in edge) {
							tg.assert(edge.graph !== undefined, "missing graph");
							let artifact = await edge.graph.get(edge.node);
							return [name, artifact];
						}
						return [name, edge];
					}),
				),
			);
		} else {
			let graph = object.graph;
			tg.assert(graph !== undefined);
			let nodes = await graph.nodes();
			let node = nodes[object.node];
			tg.assert(
				node !== undefined && node.kind === "directory",
				"expected a directory",
			);
			entries = Object.fromEntries(
				await Promise.all(
					Object.entries(node.entries).map(async ([name, edge]) => {
						if (typeof edge === "number") {
							let artifact = await graph.get(edge);
							return [name, artifact];
						} else if ("node" in edge) {
							let artifact = await (edge.graph ?? graph).get(edge.node);
							return [name, artifact];
						}
						return [name, edge];
					}),
				),
			);
		}
		tg.assert(entries !== undefined);
		for (let [name, artifact] of Object.entries(entries)) {
			yield [name, artifact];
		}
	}
}

export namespace Directory {
	export type Id = string;

	export type State = tg.Object.State<tg.Directory.Id, tg.Directory.Object>;

	export type Arg = undefined | tg.Directory | tg.Directory.Arg.Object;

	export namespace Arg {
		export type Object =
			| tg.Graph.Arg.Reference
			| {
					[key: string]:
						| undefined
						| string
						| Uint8Array
						| tg.Blob
						| tg.Artifact
						| tg.Directory.Arg.Object;
			  };
	}

	export type Object = tg.Graph.Reference | tg.Graph.Directory;

	export namespace Object {
		export let toData = (object: tg.Directory.Object): tg.Directory.Data => {
			if ("node" in object) {
				return tg.Graph.Reference.toData(object);
			} else {
				return tg.Graph.Directory.toData(object);
			}
		};

		export let fromData = (data: tg.Directory.Data): tg.Directory.Object => {
			if (tg.Graph.Data.Reference.is(data)) {
				return tg.Graph.Reference.fromData(data);
			} else {
				return tg.Graph.Directory.fromData(data);
			}
		};

		export let children = (object: tg.Directory.Object): Array<tg.Object> => {
			if ("node" in object) {
				return tg.Graph.Reference.children(object);
			} else {
				return tg.Graph.Directory.children(object);
			}
		};
	}

	export type Data = tg.Graph.Data.Reference | tg.Graph.Data.Directory;
}

File: packages/clients/python/tangram/directory.py
"""Directory class for Tangram.

This module mirrors the JavaScript packages/clients/js/src/directory.ts.
"""

from __future__ import annotations

from typing import (
    TYPE_CHECKING,
    Any,
    AsyncIterator,
    Dict,
    List,
    Optional,
    Tuple,
    Union,
)

from tangram.object_ import Object_

if TYPE_CHECKING:
    from tangram.file import File
    from tangram.symlink import Symlink

Artifact = Union["Directory", "File", "Symlink"]


async def directory(*args: Any) -> "Directory":
    """Create a new directory."""
    return await Directory.new(*args)


class Directory:
    """A directory is an artifact containing a map of names to artifacts."""

    _state: Object_.State

    def __init__(
        self,
        *,
        id: Optional[str] = None,
        object: Optional[Dict[str, Any]] = None,
        stored: bool = False,
    ) -> None:
        """Initialize a directory."""
        obj = {"kind": "directory", "value": object} if object is not None else None
        self._state = Object_.State(id=id, object=obj, stored=stored)

    @property
    def state(self) -> Object_.State:
        """Get the object state."""
        return self._state

    @classmethod
    def with_id(cls, id: str) -> "Directory":
        """Create a directory with the given ID."""
        return cls(id=id, stored=True)

    @classmethod
    def with_object(cls, object: Dict[str, Any]) -> "Directory":
        """Create a directory with the given object."""
        return cls(object=object, stored=False)

    @classmethod
    def from_data(cls, data: Dict[str, Any]) -> "Directory":
        """Create a directory from serialized data."""
        return cls.with_object(Directory.Object.from_data(data))

    @classmethod
    async def new(cls, *args: Any) -> "Directory":
        """Create a new directory from arguments."""
        import tangram as tg

        resolved = [await tg.resolve(arg) for arg in args]
        entries: Dict[str, Artifact] = {}

        for arg in resolved:
            if arg is None:
                continue
            elif isinstance(arg, Directory):
                for name, entry in (await arg.entries()).items():
                    existing_entry = entries.get(name)
                    if isinstance(existing_entry, Directory) and isinstance(entry, Directory):
                        entry = await Directory.new(existing_entry, entry)
                    entries[name] = entry
            elif isinstance(arg, dict):
                for key, value in arg.items():
                    components = _path_components(key)
                    if not components:
                        raise RuntimeError("the path must have at least one component")
                    first_component = components[0]
                    trailing_components = components[1:]

                    if first_component in (".", "..") or first_component.startswith("/"):
                        raise RuntimeError("all path components must be normal")
                    name = first_component

                    existing_entry = entries.get(name)
                    if not isinstance(existing_entry, Directory):
                        existing_entry = None

                    if trailing_components:
                        trailing_path = "/".join(trailing_components)
                        new_entry = await Directory.new(existing_entry, {trailing_path: value})
                        entries[name] = new_entry
                    else:
                        if value is None:
                            entries.pop(name, None)
                        elif isinstance(value, (str, bytes)):
                            new_entry = await tg.file(value)
                            entries[name] = new_entry
                        elif isinstance(value, tg.Blob):
                            new_entry = await tg.file(value)
                            entries[name] = new_entry
                        elif isinstance(value, (tg.File, tg.Symlink)):
                            entries[name] = value
                        elif isinstance(value, Directory):
                            if existing_entry is not None:
                                value = await Directory.new(existing_entry, value)
                            entries[name] = value
                        elif isinstance(value, dict):
                            entries[name] = await Directory.new(existing_entry, value)
                        else:
                            raise RuntimeError(f"unsupported value type: {type(value)}")

        return cls.with_object({"entries": entries})

    @staticmethod
    def expect(value: Any) -> "Directory":
        """Assert that value is a Directory and return it."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, Directory))
        return value

    @staticmethod
    def assert_(value: Any) -> None:
        """Assert that value is a Directory."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, Directory))

    @property
    def id(self) -> str:
        """Get the directory ID."""
        from tangram.assert_ import assert_

        id = self._state.id
        assert_(Object_.Id.kind(id) == "directory")
        return id

    async def object(self) -> Dict[str, Any]:
        """Get the directory object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "directory")
        return obj.get("value", obj)

    async def load(self) -> Dict[str, Any]:
        """Load the directory object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "directory")
        return obj.get("value", obj)

    def unload(self) -> None:
        """Unload the directory data."""
        self._state.unload()

    async def store(self) -> str:
        """Store the directory and return its ID."""
        import tangram as tg

        await tg.Value.store(self)
        return self.id

    async def children(self) -> List[Object_]:
        """Get child objects."""
        return await self._state.children()

    async def get(self, path: str) -> Artifact:
        """Get an artifact at the given path."""
        from tangram.assert_ import assert_

        artifact = await self.try_get(path)
        assert_(artifact is not None, f'Failed to get the directory entry "{path}".')
        return artifact

    async def try_get(self, path: str) -> Optional[Artifact]:
        """Get an artifact at the given path, returning None if not found."""
        import tangram as tg

        components = _path_components(path)
        artifact: Artifact = self
        parents: List[Directory] = []

        while True:
            if not components:
                break
            component = components.pop(0)
            if component == "/":
                raise RuntimeError("invalid path")
            elif component == ".":
                continue
            elif component == "..":
                if not parents:
                    raise RuntimeError("path is external")
                artifact = parents.pop()
                continue

            if not isinstance(artifact, Directory):
                return None

            entries = await artifact.entries()
            entry = entries.get(component)
            if entry is None:
                return None

            parents.append(artifact)
            artifact = entry

            if isinstance(entry, tg.Symlink):
                artifact_ = await entry.artifact()
                path_ = await entry.path()
                if artifact_ is None and path_ is not None:
                    if not parents:
                        raise RuntimeError("path is external")
                    artifact = parents.pop()
                    components = _path_components(path_) + components
                elif artifact_ is not None and path_ is None:
                    return artifact_
                elif isinstance(artifact_, Directory) and path_ is not None:
                    return await artifact_.try_get(path_)
                else:
                    raise RuntimeError("invalid symlink")

        return artifact

    async def entries(self) -> Dict[str, Artifact]:
        """Get all directory entries."""
        import tangram as tg

        result: Dict[str, Artifact] = {}
        async for name, artifact in self:
            result[name] = artifact
        return result

    async def walk(self) -> AsyncIterator[Tuple[str, Artifact]]:
        """Recursively walk directory yielding (path, artifact) tuples."""
        async for name, artifact in self:
            yield name, artifact
            if isinstance(artifact, Directory):
                async for subpath, subartifact in artifact.walk():
                    yield f"{name}/{subpath}", subartifact

    async def __aiter__(self) -> AsyncIterator[Tuple[str, Artifact]]:
        """Iterate over directory entries asynchronously."""
        import tangram as tg
        from tangram.assert_ import assert_

        obj = await self.object()
        entries_data = obj.get("entries", {})
        for name, entry in entries_data.items():
            if isinstance(entry, Object_):
                yield name, entry
            elif isinstance(entry, str):
                artifact = _artifact_with_id(entry)
                yield name, artifact
            elif isinstance(entry, dict) and "index" in entry:
                graph = entry.get("graph")
                assert_(graph is not None, "missing graph")
                artifact = await graph.get(entry["index"])
                yield name, artifact
            else:
                yield name, entry


def _path_components(path: str) -> List[str]:
    """Split a path into components."""
    if not path:
        return []
    components = []
    if path.startswith("/"):
        components.append("/")
        path = path[1:]
    parts = path.split("/")
    for part in parts:
        if part and part != ".":
            components.append(part)
    return components


def _artifact_with_id(id: str) -> Artifact:
    """Create an artifact from its ID."""
    import tangram as tg

    prefix = id[:3] if len(id) >= 3 else ""
    if prefix == "dir":
        return Directory.with_id(id)
    elif prefix == "fil":
        return tg.File.with_id(id)
    elif prefix == "sym":
        return tg.Symlink.with_id(id)
    else:
        raise ValueError(f"invalid artifact id: {id}")


# Nested Object namespace.
class _DirectoryObject:
    """Namespace for Directory.Object operations."""

    @staticmethod
    def to_data(obj: Dict[str, Any]) -> Dict[str, Any]:
        """Convert object to data."""
        data: Dict[str, Any] = {}
        entries = obj.get("entries", {})
        if entries:
            data["entries"] = {}
            for name, artifact in entries.items():
                if hasattr(artifact, "id"):
                    data["entries"][name] = artifact.id
                else:
                    data["entries"][name] = artifact
        return data

    @staticmethod
    def from_data(data: Dict[str, Any]) -> Dict[str, Any]:
        """Convert data to object."""
        obj: Dict[str, Any] = {}
        entries_data = data.get("entries", {})
        entries = {}
        for name, entry_data in entries_data.items():
            if isinstance(entry_data, str):
                entries[name] = _artifact_with_id(entry_data)
            else:
                entries[name] = entry_data
        obj["entries"] = entries
        return obj

    @staticmethod
    def children(obj: Optional[Dict[str, Any]]) -> List[Object_]:
        """Get child objects from directory object."""
        if obj is None:
            return []
        result: List[Object_] = []
        entries = obj.get("entries", {})
        for artifact in entries.values():
            if isinstance(artifact, Object_):
                result.append(artifact)
        return result


Directory.Object = _DirectoryObject

# Type aliases.
Directory.Id = str
Directory.Arg = Union[None, "Directory", Dict[str, Any]]
Directory.Data = Dict[str, Any]
